File: Java/Problem4.java
class Solution {
    public String shortestPalindrome(String s) {
        int p = 31;
        int mod = (int) 1e9 + 7;
        int length = s.length();

        long[] powers = new long[length];
        long currentPower = 1;

        for (int i = 0; i < length; i++) {
            powers[i] = currentPower;
            currentPower = (currentPower * p) % mod;
        }

        long[] hashesLeft = new long[length];
        long[] hashesRight = new long[length];

        for (int i = 0; i < length; i++) {
            hashesLeft[i] = ((s.charAt(i) - 'a' + 1) * powers[i]) % mod;
            hashesRight[length - i - 1] = ((s.charAt(length - i - 1) - 'a' + 1) * powers[i]) % mod;

            if (i > 0) {
                hashesLeft[i] = (hashesLeft[i] + hashesLeft[i - 1]) % mod;
                hashesRight[length - i - 1] = (hashesRight[length - i - 1] + hashesRight[length - i]) % mod;
            }
        }

        int k;
        for (k = 0; k < length; k++) {
            long hashPrefix = hashesLeft[length - k - 1];
            long hashInvSuffix = hashesRight[0];

            if (k > 0) {
                hashPrefix = (hashPrefix * powers[k]) % mod;
                hashInvSuffix = (hashInvSuffix - hashesRight[length - k] + mod) % mod;
            }

            if (hashPrefix == hashInvSuffix) {
                boolean isPalindrome = true;
                for (int i = 0; i < (length - k) / 2; i++) {
                    if (s.charAt(i) != s.charAt(length - k - i - 1)) {
                        isPalindrome = false;
                        break;
                    }
                }
                if (isPalindrome) break;
            }
        }

        StringBuilder result = new StringBuilder(s.substring(length - k)).reverse();
        result.append(s);

        return result.toString();
    }
}

File: Python/Problem4.py
class Solution:
    def shortestPalindrome(self, s: str) -> str:
        p = 31
        mod = int(1e9 + 7)
        length = len(s)

        powers = [1] * length
        for i in range(1, length):
            powers[i] = (powers[i - 1] * p) % mod

        hashes_left = [0] * length
        hashes_right = [0] * length

        for i in range(length):
            hashes_left[i] = ((ord(s[i]) - ord('a') + 1) * powers[i]) % mod
            hashes_right[length - i - 1] = ((ord(s[length - i - 1]) - ord('a') + 1) * powers[i]) % mod

            if i > 0:
                hashes_left[i] = (hashes_left[i] + hashes_left[i - 1]) % mod
                hashes_right[length - i - 1] = (hashes_right[length - i - 1] + hashes_right[length - i]) % mod

        k = 0
        for k in range(length):
            hash_prefix = hashes_left[length - k - 1]
            hash_inv_suffix = hashes_right[0]

            if k > 0:
                hash_prefix = (hash_prefix * powers[k]) % mod
                hash_inv_suffix = (hash_inv_suffix - hashes_right[length - k] + mod) % mod

            if hash_prefix == hash_inv_suffix:
                is_palindrome = True
                for i in range((length - k) // 2):
                    if s[i] != s[length - k - i - 1]:
                        is_palindrome = False
                        break
                if is_palindrome:
                    break

        result = s[length - k:][::-1] + s 
        return result

        